### PR TITLE
Add dehydrated flag to device_keys of dehydrated devices

### PR DESCRIPTION
**Pull request by uhoreg**

### CODE REVIEW

File: crates/matrix-sdk-crypto/src/dehydrated_devices.rs
Comment (poljar) [inline crates/matrix-sdk-crypto/src/dehydrated_devices.rs:321]: This should happen in the `DehydratedDevices::create()` method.

File: crates/matrix-sdk-crypto/src/olm/account.rs
Comment (poljar) [inline crates/matrix-sdk-crypto/src/olm/account.rs:435]: I think that this should be a constructor. We don't want to have the ability to convert an existing account into a `Dehydrated` one, we want to either create a normal `Account` or for dehydration.

The `DehydratedDevices::create()` method then creates a dehydrated one.

**Review by poljar — CHANGES_REQUESTED** [review]
I know this is still a draft, but since the PR is small I took a look. Left some comments.

File: crates/matrix-sdk-crypto/src/dehydrated_devices.rs
Comment (uhoreg) [reply crates/matrix-sdk-crypto/src/dehydrated_devices.rs:321]: Yup.  I just put this here for now because I didn't want to touch the storage parts yet (which is also why I didn't put it in a constructor).

File: crates/matrix-sdk-crypto/src/olm/account.rs
Comment (uhoreg) [inline crates/matrix-sdk-crypto/src/olm/account.rs:363]: adding this flag here will break things that create PickledAccount, such as https://github.com/matrix-org/matrix-rust-sdk-crypto-wasm/blob/main/src/libolm_migration.rs#L136  I'm not sure if there's a good way to deal with this.

Comment (poljar) [reply crates/matrix-sdk-crypto/src/olm/account.rs:363]: They'll need to update the bindings just like you did with the `matrix-sdk-crypto-ffi` bindings.

File: crates/matrix-sdk-crypto/src/dehydrated_devices.rs
Comment (poljar) [inline crates/matrix-sdk-crypto/src/dehydrated_devices.rs:483]:
```suggestion
        assert_eq!(device_keys.dehydrated, Some(true), "The device keys of the dehydrated device should be marked as dehydrated.");
```

File: crates/matrix-sdk-crypto/src/dehydrated_devices.rs
Comment (poljar) [inline crates/matrix-sdk-crypto/src/dehydrated_devices.rs:481]: Doesn't `request.device_keys.deserialize_as()` work here? Also please import `DeviceKeys` if you need them.

File: crates/matrix-sdk-crypto/src/identities/device.rs
Comment (poljar) [inline crates/matrix-sdk-crypto/src/identities/device.rs:485]:
```suggestion
    /// Whether or not the device is a dehydrated device.
```

File: crates/matrix-sdk-crypto/src/olm/account.rs
Comment (poljar) [inline crates/matrix-sdk-crypto/src/olm/account.rs:164]:
```suggestion
    /// Whether the account is for a dehydrated device.
```

**Review by poljar — CHANGES_REQUESTED** [review]
Left some small nits, looks generally good.

File: crates/matrix-sdk-crypto/src/olm/account.rs
Comment (uhoreg) [reply crates/matrix-sdk-crypto/src/olm/account.rs:363]: OK, I just wasn't sure if there was a better way to do it.  I assume this means that the version will need to be bumped since this is a breaking change.  Is there a way to flag this so that when the next release gets made, the version will be bumped correctly?

Comment (poljar) [reply crates/matrix-sdk-crypto/src/olm/account.rs:363]: We're still in the `0.X.Y` range of releases, so the correct version will get bumped by design. Adding this as a breaking change to the changelog might make sense to notify people about it.

File: crates/matrix-sdk-crypto/src/dehydrated_devices.rs
Comment (uhoreg) [reply crates/matrix-sdk-crypto/src/dehydrated_devices.rs]: changed to use `deserialize_as` (I didn't know about the function).  `mod tests` already imports `ruma::encryption::DeviceKeys`, so I can't import `crate::types::DeviceKeys` at the module-level.  Do I just `use crate::types::DeviceKeys` at the top of this function?  It looks like that's possible, but I've never seen that done, so I don't know if that's a "normal" thing to do.

Comment (poljar) [reply crates/matrix-sdk-crypto/src/dehydrated_devices.rs]: It's possible, you can also rename things if you import things using `use Foo as Bar`.

**Review by poljar — APPROVED** [review]
Looks good. Could you get rid of the merge conflict so CI kicks in. After that we can merge.

File: crates/matrix-sdk-crypto/src/olm/account.rs
Comment (uhoreg) [inline crates/matrix-sdk-crypto/src/olm/account.rs:441]: I meant to ask earlier, but on the MSC, I suggested using URL-safe unpadded Base64 instead of plain Base64, since the device ID gets put in path parameters.  If you don't have any objections to that, I may as well change this code to make this device ID URL-safe before we merge.

Comment (poljar) [reply crates/matrix-sdk-crypto/src/olm/account.rs:441]: We were pondering this when we did the iOS implementation and decided against it. The reason being that it would be confusing to have two different representations for a Curve25510 key. One being in the path and one being uploaded as the device keys. Both encodings would end up in the logs as well and confuse potential bug hunters.

File: crates/matrix-sdk-crypto/CHANGELOG.md
Comment (poljar) [inline crates/matrix-sdk-crypto/CHANGELOG.md:19]: I don't think that this belongs here :sweat_smile:

**Review by poljar — CHANGES_REQUESTED** [review]

**Review by poljar — APPROVED** [review]
Looks good, just ensure to make CI happy with the formatting.

**Comment by uhoreg** [comment]
Code coverage failed due to hitting a limit with codecov.io.  But the previous commit was OK, and the last commit was just a formatting change, so should be OK.

I don't have permissions in this repo, so someone will need to hit the merge button for me.

**Pull request merged**
